### PR TITLE
[vLLM] Expand and optimize CPU sampling for non-greedy serving 

### LINKED ISSUE #4370: [vLLM] Expand and optimize CPU sampling for non-greedy serving
## Summary

Device-side non-greedy sampling is currently ~5 tok/s  (WIP https://github.com/tenstorrent/tt-xla/issues/3940 improvements coming) -- unusable for demos or production. Device sampling performance is being actively worked on separately, but in the near term `cpu_sampling=True` is the practical workaround. We've extended the CPU sampling path to support full non-greedy sampling (temperature, penalties, top-k/top-p) and optimized it with the Gumbel-max trick. These changes have been running in demos for the past 3 weeks.

Note: we also had a fix for the cpu_sampling routing bug (path never being reached) but dropped it since main already has the same fix via https://github.com/tenstorrent/tt-xla/pull/4332.

Planning to put up a PR from branch `kmabee/cpu_sampling_improvements`. The relevant commits:

| Commit | Description |
|---|---|
| `3f215c795` | Extend CPU sampling to support temperature, penalties, and top-k/top-p |
| `7f3848aae` | Optimize CPU sampling: Gumbel-max trick + batched top-k path |

## Performance

Llama-3.1-8B-Instruct, single P150, cpu_sampling=True, bfp_bf8, opt_level=1:

| Batch | Forward-pass ceiling | CPU sampling | Gap |
|---|---|---|---|
| 2 (demo) | 21.7 tok/s | 19.0 tok/s | 12% |
| 8 | 19.7 tok/s | 15.1 tok/s | 23% |

The remaining gap is the device-to-CPU logits transfer. The main optimization was replacing `torch.softmax + torch.multinomial` with the Gumbel-max trick (`argmax(logits + Gumbel noise)`) -- mathematically equivalent but avoids `multinomial` which processes each batch element sequentially.

## Gumbel-max optimization

The main CPU sampling bottleneck was `torch.multinomial` -- it processes each batch element sequentially and is expensive at vocab size 128K. We replaced it with the Gumbel-max trick: adding Gumbel-distributed noise to the logits and taking `argmax`. This is a well-known method for sampling from a categorical distribution without explicitly computing softmax or calling multinomial -- `argmax(logits + Gumbel(0,1))` produces samples from exactly the same distribution as `multinomial(softmax(logits))`.

```python
# Before: softmax + multinomial (~4x slower at vocab=128K)
probs = torch.softmax(logits, dim=-1)
random = torch.multinomial(probs, num_samples=1).squeeze(-1)

# After: Gumbel-max (single batched argmax, same distribution)
gumbel = -torch.log(-torch.log(torch.rand_like(logits.float()) + 1e-20) + 1e-20)
random = torch.argmax(logits + gumbel, dim=-1)
```

This matches what the device-side sampler already does -- `random_sample()` in `sampler.py` uses the equivalent exponential trick (`probs.div_(exponential).argmax()`). Dividing probabilities by Exp(1) random variables and taking argmax is mathematically the same as adding Gumbel noise to log-probabilities and taking argmax. Both avoid `torch.multinomial`.

### PR DESCRIPTION
**Pull request by kmabeeTT**
## Ticket

Closes #4370

## Problem

- `sample_from_logits_cpu` only supported greedy decoding. Device-side non-greedy sampling is currently ~5 tok/s due to and (improvements WIP, but more involved) so `cpu_sampling=True` is the practical path for non-greedy serving that was used for tt-cloud-console demo last 3 weeks.  Merging these changes since they are small and contained while device sampling is improved.

## What's changed

- `model_runner.py`: extend `sample_from_logits_cpu` to support temperature scaling, repetition/frequency/presence penalties, top-k, and top-p
- `model_runner.py`: replace `torch.softmax + torch.multinomial` with the Gumbel-max trick — `argmax(logits + Gumbel noise)` is mathematically equivalent to `multinomial(softmax(logits))` but avoids `multinomial` which processes each batch element sequentially
- `model_runner.py`: add batched top-k fast path — when all requests share the same `top_k > 0`, a single `torch.topk` call reduces vocab from 128K to k elements before top-p and Gumbel-max, eliminating per-request loops (see details in ticket).
- `test_cpu_sampling.py` (new): correctness tests for cpu_sampling covering greedy, temperature, top-k, top-p, and repetition penalty using Llama-3.2-1B-Instruct

## Checklist

- [x] Non-greedy output coherent on Llama-3.1-8B at batch 2 (19.0 tok/s, 3 weeks of demo use)
- [x] Greedy throughput unchanged
- [x] `test_cpu_sampling.py` passing
- [x] Run vllm nightly on CI passing: https://github.com/tenstorrent/tt-xla/actions/runs/24802050427

### CODE REVIEW

**Review by AleksKnezevic — APPROVED** [review]

**Comment by kmabeeTT** [comment]
As further test, dropped changes in tt-xla checkout that powers the tt-inference-server, cherry pick these and the previous fix in main:

```
2026-04-22 (b6a0e7dc1) by Kyle Mabee (Author kmabee@tenstorrent.com) : Optimize CPU sampling: Gumbel-max trick + batched top-k path
2026-04-22 (700791307) by Kyle Mabee (Author kmabee@tenstorrent.com) : Extend CPU sampling to support temperature, penalties, and top-k/top-p
2026-04-22 (fa5304ba3) by Kyle Mabee (Author mmanzoor@tenstorrent.com) : [vLLM] Fix cpu sampling (#4332)
```

still works:

```
Prompt (q to quit): Okay cool, keep going with that story.
Response: As the stranger, Marcus, entered the dimly lit shop, the scent of polished wood and oil wafted through the air, mingling with the faint tang of metal shavings. Leo, hunched over a workbench, looked up from his latest creation, a delicate pocket watch with an intricately engraved face
[64 tokens | TTFT: 89ms | 3.50s | 18.3 tok/s | in:273 out:65]

Prompt (q to quit):
```

(slightly lower than 19 tok/sc, since disabled trace here in this experiment).

File: integrations/vllm_plugin/vllm_tt/model_runner.py [hunk 1]
```diff
@@ -2167,13 +2177,96 @@ def sample_from_logits_cpu(
...
+        Supports greedy, temperature, top-k/top-p, and penalty-based sampling.
+        All operations run on CPU to avoid compiling a device sampling graph.
         """
-        device = logits.device
         logits = logits.cpu()
-        out_tokens = torch.argmax(logits, dim=-1, keepdim=True)
-        return out_tokens.to(device)
+
+        if not sampling_metadata.no_penalties:
+            output_counts = sampling_metadata.output_token_counts.cpu()
```
Comment (mmanzoorTT) [inline integrations/vllm_plugin/vllm_tt/model_runner.py]: We are creating `sampling_metadata` on device if `cpu-sampling` is enabled. I don't see a need to use `.cpu()`. These variables should work without moving to `cpu`.

**Review by mmanzoorTT — APPROVED** [review]

File: integrations/vllm_plugin/vllm_tt/model_runner.py (hunk 1 above)
Comment (kmabeeTT) [reply integrations/vllm_plugin/vllm_tt/model_runner.py]: Good point - just updated with new commit for this line and all others. Tested it out on inference server quickly, still works.
